qt-build/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+};
+
+pub use versions::SemVer;
+
+/// Helper for build.rs scripts using Qt
+/// ```
+/// let qt_modules = vec!["Core", "Gui"]
+///     .iter()
+///     .map(|m| String::from(*m))
+///     .collect();
+/// let qtbuild = qt_build::QtBuild::new(qt_modules).expect("Could not find Qt installation");
+/// ```
+pub struct QtBuild {
+    pub version: SemVer,
+    qmake_executable: String,
+    qt_modules: Vec<String>,
+}
+
+impl QtBuild {
+    /// Search for where Qt is installed using qmake. Specify the Qt modules you are
+    /// linking with the `qt_modules` parameter, ommitting the `Qt` prefix (`"Core"`
+    /// rather than `"QtCore"`). After construction, use the [QtBuild::qmake_query]
+    /// method to get information about the Qt installation.
+    ///
+    /// The directories specified by the `PATH` environment variable are where qmake is
+    /// searched for. Alternatively, the `QMAKE` environment variable may be set to specify
+    /// an explicit path to qmake.
+    ///
+    /// If multiple major versions (for example, `5` and `6`) of Qt could be installed, set
+    /// the `QT_VERSION_MAJOR` environment variable to force which one to use. When using Cargo
+    /// as the build system for the whole build, prefer using `QT_VERSION_MAJOR` over the `QMAKE`
+    /// environment variable because it will account for different names for the qmake executable
+    /// that some Linux distributions use.
+    ///
+    /// However, when building a Rust staticlib that gets linked to C++ code by a C++ build
+    /// system, it is best to use the `QMAKE` environment variable to ensure that the Rust
+    /// staticlib is linked to the same installation of Qt that the C++ build system has
+    /// detected. With CMake, you can get this from the `Qt${QT_VERSION_MAJOR}::qmake`
+    /// target's `IMPORTED_LOCATION` property, for example:
+    /// ```
+    /// find_package(QT NAMES Qt6 Qt5 COMPONENTS Core REQUIRED)
+    /// find_package(Qt${QT_VERSION_MAJOR} COMPONENTS Core REQUIRED)
+    /// get_target_property(QMAKE Qt${QT_VERSION_MAJOR}::qmake IMPORTED_LOCATION)
+    ///
+    /// execute_process(
+    ///     COMMAND cmake -E env
+    ///         "CARGO_TARGET_DIR=${CMAKE_CURRENT_BINARY_DIR}/cargo"
+    ///         "QMAKE=${QMAKE}"
+    ///         cargo build
+    ///     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    /// )
+    /// ```
+    #[allow(clippy::result_unit_err)]
+    pub fn new(mut qt_modules: Vec<String>) -> Result<Self, ()> {
+        if qt_modules.is_empty() {
+            qt_modules.push("Core".to_string());
+        }
+        println!("cargo:rerun-if-env-changed=QMAKE");
+        println!("cargo:rerun-if-env-changed=QT_VERSION_MAJOR");
+        fn verify_candidate(candidate: &str) -> Result<(&str, versions::SemVer), String> {
+            match Command::new(&candidate)
+                .args(&["-query", "QT_VERSION"])
+                .output()
+            {
+                Err(_) => Err("qmake -query failed".to_string()),
```

Comment:
So could even have these inner QMake errors be their own Enum and then nest that into the outer enum or share the same one.